### DISCUSSION TITLE
pnpm v7

### DISCUSSION THREAD

**Discussion by zkochan** [pnpm/pnpm#3536] [General]
This is a discussion about pnpm v7 breaking changes.

- ~The biggest breaking change will be that pnpm will be shipped as a compiled binary, so pnpm won't need Node.js preinstalled on a system. This version of pnpm is already shipped as [@pnpm/beta](https://www.npmjs.com/package/@pnpm/beta)~
- [x] The pnpx command will be removed. Instead of pnpx, pnpm may be used. Can we remove it now that Corepack ships  it? Maybe just make it an alias of `pnpm dlx`.
- ~Turn off hoisting by default (at least in monorepos)~
- [x] Turn on strict-peer-dependencies by default
- [x] Exclude the root from recursive script runner commands: https://github.com/pnpm/pnpm/issues/2769
- [x] `pnpm pack` should only pack a file as an executable, if it is a bin or listed in the `executableFiles` array
- [x] Remove the `-W` option. `--workspace-root/-w` should be used instead.
- [x] `extend-node-path` should be `false` by default. The setting should be removed.
- ~rename `use-node-version` to `node-version`.~
- [x] remove npm_config_argv https://github.com/pnpm/pnpm/discussions/4153
- [x] make [`side-effects-cache`](https://pnpm.io/npmrc#side-effects-cache) true by default.
- [x] make `embed-readme` false by default
- [x] Drop Node.js 12 support (EOL will be 2022-04-30)
- [x] `pnpm install -g pkg` write bin commands only to predefined location.
- [x] When using `pnpm run <script>`, all command line arguments after the script name are passed to the script's argv. https://github.com/pnpm/pnpm/issues/3778
- [x] The file protocol should work the same as in Yarn: it should copy the files from the package, not symlink the package.
- ~Install dev deps even if NODE_ENV is set to `production`.~
- [x] Improve how beta versions are handled
- [x] `pnpm init` is non-interactive by default.
- [x] pnpm store location should be inside the pnpm home directory

Review the breaking changes in Yarn v4 and do the same changes, when it makes sense: https://github.com/yarnpkg/berry/issues/3591

Related PRs:
- https://github.com/pnpm/pnpm/pull/4253
- https://github.com/pnpm/pnpm/pull/4266
- https://github.com/pnpm/pnpm/pull/4267
- https://github.com/pnpm/pnpm/pull/4290
- https://github.com/pnpm/pnpm/pull/4280
- https://github.com/pnpm/pnpm/pull/4307
- https://github.com/pnpm/pnpm/pull/4308
- https://github.com/pnpm/pnpm/pull/4408
- https://github.com/pnpm/pnpm/pull/4416
- https://github.com/pnpm/pnpm/pull/4427
- https://github.com/pnpm/pnpm/pull/4435
- https://github.com/pnpm/pnpm/pull/4469
- https://github.com/pnpm/pnpm/pull/4513
- https://github.com/pnpm/pnpm/pull/4519
- https://github.com/pnpm/pnpm/pull/4522

**Comment by ibezkrovnyi** [comment]
Few thoughts:
1. Shipping pnpm as a compiled binary while we already have node binaries downloaded will require an additional 36 MB downloaded per each GitLab pipeline job. 20 jobs = 720 MBs of unnecessary downloads on each commit.
5. One of the biggest advantages of pnpm is the ability to work with simple repo as with monorepo with a single project (no matter if pnpm-workspace.yaml is present or not) . With using GitLab CI pipeline configured for pnpm monorepo commands `pnpm -r --filter <FILTER> run <SCRIPT>`, `pnpm -r --filter <FILTER> exec <CMD>` and `pnpm -r --filter <FILTER> ls --depth -1 --json` will work with one root-level project just fine. So, currently there is no need to have separate GitLab CI pipelines for pnpm monorepo and for simple repo. It would be nice to keep this pnpm's superpower (at least with some `.npmrc` option).

**Reply by zkochan** [reply]
> Shipping pnpm as a compiled binary while we already have node binaries downloaded will require an additional 36 MB downloaded per each GitLab pipeline job. 20 jobs = 720 MBs of unnecessary downloads on each commit.

We will probably keep pnpm to be js but we will recommend to use the packaged version, which is faster.

Can't you create something like a base image for your pipeline with pnpm preinstalled?

> One of the biggest advantages of pnpm is the ability to work with simple repo as with monorepo with a single project (no matter if pnpm-workspace.yaml is present or not) . With using GitLab CI pipeline configured for pnpm monorepo commands pnpm -r --filter <FILTER> run <SCRIPT>, pnpm -r --filter <FILTER> exec <CMD> and pnpm -r --filter <FILTER> ls --depth -1 --json will work with one root-level project just fine. So, currently there is no need to have separate GitLab CI pipelines for pnpm monorepo and for simple repo. It would be nice to keep this pnpm's superpower (at least with some .npmrc option).

Which of the items break this workflow?

**Reply by ibezkrovnyi** [reply]
> Can't you create something like a base image for your pipeline with pnpm preinstalled?

Yes, we can preinstall pnpm binary into docker image, so the size of the image will grow. Docker image is downloaded in each GitLab job, so there is no gain preinstalling pnpm.

> Which of the items break this workflow?

I was under impression that this one: `Exclude the root from recursive script runner commands`, but after reading related PR https://github.com/pnpm/pnpm/pull/3647 I think the only change that will be needed to make pnpm treating monorepo and single repo, in the same way, is to add `--workspace-root` (`-w`), so this is not a problem then.

**Comment by donferi** [comment]
One question, will pnpm shipped as a compiled binary work fine with `corepack`?

**Reply by zkochan** [reply]
We will ship both the compiled binary and the js version. The main pnpm package will stay js.

**Reply by Jack-Works** [reply]
since you must use pnpm with Node (you didn't RIIR right?) why it will bring its own Node.js runtime? It's a bit strange.

**Reply by zkochan** [reply]
It is packaged by [pkg](https://github.com/vercel/pkg).

It allows to use pnpm without Node.js preinstalled. So you can install Node.js with pnpm. 

It also makes pnpm faster.

**Reply by rxliuli** [reply]
This is indeed a bit strange, it seems that if you don't rewrite it in other languages, you don't need to bind nodejs. After all, what is the need for pnpm but not nodejs? And it also makes it difficult to upgrade (at least on windows)

**Reply by zkochan** [reply]
it is strange. But don't you prefer to have a faster pnpm?

also, many users like this feature and they use pnpm now instead of nvm to manage Node.js versions.

**Reply by zkochan** [reply]
I did not notice issues on Windows.

**Reply by Jack-Works** [reply]
> It also makes pnpm faster.

I'm curious how bringing your own Node will make pnpm faster. Does it use some kind of V8 memory snapshot like what Deno does to the TypeScript compiler?

**Reply by zkochan** [reply]
pkg compiles pnpm to v8 bytecode: https://github.com/vercel/pkg#bytecode-reproducibility

**Reply by Jack-Works** [reply]
Cool, does that has a significant performance improvement?

**Reply by gluxon** [reply]
From some tests on an Ubuntu image, I did notice that the prepackaged binary started in 0.25s vs 0.5s for the JS version. That seems small, but saving a quarter second on every pnpm invocation seems significant to me.

**Reply by zkochan** [reply]
I think it makes startup faster by 0.33 seconds on my machine.

**Comment by zkochan** [comment]
I started to work on this @pnpm/collaborators

**Comment by PabloSzx** [comment]
> Drop Node.js 12 support

Is this change intended to align with Node.js v12 EOL on 30th of April? https://nodejs.org/en/about/releases if not, I feel like this might make harder the pnpm usage on libraries that need to support Node.js v12 before that happens

> Maybe just make it an alias of pnpm dlx

I can agree with `pnpx` being an alias for `pnpm dlx`, the only issue this might have is moving away from the npm npx logic of, if the package is already available in node_modules, it uses that instead of installing it from zero

> Turn on strict-peer-dependencies by default (at least in monorepos)

This will make the life harder on any people that use React or GraphQL, just go give an example, to this day, Next.js still has peer dependency issues with the missing `@babel/core`, which is clearly not recommended to install manually.

![image](https://user-images.githubusercontent.com/8672915/150435805-7d7ebf55-f30e-4945-9413-c1da6f5745f4.png)

**Reply by zkochan** [reply]
> Is this change intended to align with Node.js v12 EOL on 30th of April?

Maybe it will be released under "next" dist-tag and only made "latest" after April 30th.

> This will make the lifer harder on any people that use React or GraphQL, just go give an example, to this day, Next.js still has peer dependency issues with the missing @babel/core, which is clearly not recommended to install manually.

The warnings/errors may be mutate with https://pnpm.io/package_json#pnpmpeerdependencyrulesignoremissing

**Reply by anomiex** [reply]
> I can agree with pnpx being an alias for pnpm dlx, the only issue this might have is moving away from the npm npx logic of, if the package is already available in node_modules, it uses that instead of installing it from zero

Yeah, we're going to have to change a bunch of our scripts to use (I guess?) `pnpm exec` instead of `pnpx` or `pnpx --no-install`.

On the plus side, not having to remember the `--no-install` everywhere to prevent it accidentally prompting to install something globally in certain edge cases (i.e. someone forgot to `pnpm install` first) might be an improvement.

**Reply by zkochan** [reply]
You can replace `pnpx` with just pnpm. `pnpm eslint` is the same as `pnpm exec eslint`.

**Comment by tjx666** [comment]
@zkochan In pnpmv6, global packages will install to node installation path.

<img width="920" alt="image" src="https://user-images.githubusercontent.com/41773861/154832235-6d1d20c8-5e58-4d45-a341-80e18979254c.png">

In v7, you need manually define the `PNPM_HOME` to install global package.

So, I have two question:

1. Which place is recommend place to set `PNPM_HOME` in MacOS? Seems to be `~/Library/pnpm`?
2. How to migrate global packages install in pnpmV6 to new defined `PNPM_HOME` in pnpmV7

**Reply by zkochan** [reply]
On MacOS we currently recommend `~/Library/pnpm`. 

https://github.com/pnpm/pnpm/blob/cf478bb8b31c23203f0a0795bbae232245a62dd4/packages/config/src/dirs.ts#L43-L62

There's no migration script. Just reinstall the global packages.

**Reply by tjx666** [reply]
For second question, I just use nvm to uninstall current node, and reinstall it.

**Reply by tjx666** [reply]
@zkochan It seems a bit annoying.

<img width="893" alt="image" src="https://user-images.githubusercontent.com/41773861/154837393-3462eca9-6a6f-4a2b-94b9-7831dab09cb9.png">

**Reply by zkochan** [reply]
I don't have a mac to test this. Should we use another directory for global pnpm packages? Does pnpm have write access to `~/.local/share/pnpm`

**Reply by tjx666** [reply]
In before screenshot, I have deleted the pnpm folder.
After I manually create the pnpm folder `~/Library/pnpm`, no permission problem...

**Reply by eshimischi** [reply]
there is 3. When you use NodeJS version managers like nvm and have several independent distributives (ex: v16.15.0, v18.0.0) how to install global packages for every nodejs version if `~/Library/pnpm` is one ?

**Reply by zkochan** [reply]
When pnpm installs a package globally, it links it to a specific Node.js version. So you may have eslint that runs on Node.js 12 and typescript that runs on Node.js 14 while your current node.js is 16. There is no need to create separate global directories for each node.js version.

**Comment by zkochan** [comment]
Looks like sometimes it causes confusion that pnpm doesn't install dev dependencies, when NODE_ENV is set to `production`. Maybe we can change this. Yarn installs all deps, when NODE_ENV is production. npm behaves the way pnpm behaves at the moment.

**Reply by shellscape** [reply]
I'm used to this behavior and have always found yarn odd. I would think that a nice message printed to console when NODE_ENV is production would be sufficient.

**Reply by zkochan** [reply]
We can do this. I hope if will reduce the amount of issues about it. There were messages about it several time in the chat. And issues were created.

**Reply by zkochan** [reply]
Done https://github.com/pnpm/pnpm/pull/4362

**Comment by zkochan** [comment]
I think one more breaking change is needed. And this one is needed for implementing a [deploy command](https://github.com/pnpm/pnpm/issues/4378).

Currently, peer dependencies are resolved from the root of the workspace as well. As a result, even when running a filtered installations, the root workspace package should be included. I suggest to change this and don't resolve peer dependencies from the root package.json.

**Reply by anomiex** [reply]
Maybe I misunderstand how `node_modules/` is intended to work, but that seems contrary to the logic of it. A peer dependency is supposed to be installed alongside the package or in a parent directory, and in most cases the workspace root will be a parent directory, so I'd expect a peer installed there to be "found".

With npm or yarn (non-pnp), node itself would find the peer installed there.

**Reply by zkochan** [reply]
It will still work because Node.js resolution algorithm is not influenced by pnpm.

**Reply by zkochan** [reply]
PR: https://github.com/pnpm/pnpm/pull/4469

**Comment by larsivi** [comment]
I have a usecase that is a different type of deploy - we have google cloud functions in our project (aka AWS lambda). This is in practice just source code that is deployed using the gcloud cli tool. It just takes the directory given, with its package.json, uploads it, install the dependencies there and tries to run it.

The issue I have then is that I need to have the workspace link resolved prior running my deploy command, but then of course without running publish. Since our source is typescript, we already build this into its own directory, with a customized .npmrc etc (private packages). What I think would be most suitable then is to have a command like

pnpm resolve[-workspace] <dir>.

Since that dir doesn't have a node_modules folder,  the command would have to use cwd for resolution (I saw if I could fake this using publish --dry-run, but couldn't as it tried to resolve the workspace links in <dir>). Or maybe there are other existing ways to make this work?

**Comment by anomiex** [comment]
After running into the usual phantom dep random version issues, I've been trying setting `hoist-pattern` to the minimal set of packages needed in our repo, which turned out to be surprisingly few. But with pnpm 7 no longer setting `NODE_PATH` in the bin stubs, I find that Webpack's `postcss-loader` can no longer find the postcss plugins. Reproduction: [test.zip](https://github.com/pnpm/pnpm/files/8624824/test.zip)

Anyone have a better workaround than listing the plugin packages in `hoist-pattern` (ugh) or manually setting `NODE_PATH` when running webpack?

**Reply by shellscape** [reply]
Honestly I've been using `shamefully-hoist` since it was introduced and haven't had any issues.

**Comment by zkochan** [comment]
I have locked this discussion because people are posting random issues here for some reason.
